superset-frontend/src/dashboard/components/FiltersBadge/selectors.ts

```diff
@@ -0,0 +1,161 @@
+// Should patch it here until the DatasourceMeta type is updated.
+type Datasource = {
+  time_grain_sqla?: [string, string][];
+  granularity?: [string, string][];
+};
+
+type Filter = {
+  chartId: number;
+  columns: { [key: string]: string | string[] };
+  scopes: { [key: string]: any };
+  labels: { [key: string]: string };
+  isDateFilter: boolean;
+  directPathToFilter: string[];
+  datasourceId: string;
+};
+
+const selectIndicatorValue = (
+  columnKey: string,
+  filter: Filter,
+  datasource: Datasource,
+): string[] => {
+  const values = filter.columns[columnKey];
+  const arrValues = Array.isArray(values) ? values : [values];
+
+  if (
+    isNil(values) ||
+    (filter.isDateFilter && values === 'No filter') ||
+    arrValues.length === 0
+  ) {
+    return [];
+  }
+
+  if (filter.isDateFilter && TIME_GRANULARITY_FIELDS.has(columnKey)) {
+    const timeGranularityMap = (
+      (columnKey === TIME_FILTER_MAP.time_grain_sqla
+        ? datasource.time_grain_sqla
+        : datasource.granularity) || []
+    ).reduce(
+      (map, [key, value]) => ({
+        ...map,
+        [key]: value,
+      }),
+      {},
+    );
+
+    return arrValues.map(value => timeGranularityMap[value] || value);
+  }
+
+  return arrValues;
+};
+
+const selectIndicatorsForChartFromFilter = (
+  chartId: number,
+  filter: Filter,
+  filterDataSource: Datasource,
+  appliedColumns: Set<string>,
+  rejectedColumns: Set<string>,
+): Indicator[] => {
+  // filters can be applied (if the filter is compatible with the datasource)
+  // or rejected (if the filter is incompatible)
+  // or the status can be unknown (if the filter has calculated parameters that we can't analyze)
+  const getStatus = (column: string) => {
+    if (appliedColumns.has(column)) return IndicatorStatus.Applied;
+    if (rejectedColumns.has(column)) return IndicatorStatus.Incompatible;
+    return IndicatorStatus.Unset;
+  };
+
+  return Object.keys(filter.columns)
+    .filter(column =>
+      getChartIdsInFilterScope({
+        filterScope: filter.scopes[column],
+      }).includes(chartId),
+    )
+    .map(column => ({
+      column,
+      name: filter.labels[column] || column,
+      value: selectIndicatorValue(column, filter, filterDataSource),
+      status: getStatus(column),
+      path: filter.directPathToFilter,
+    }));
+};
+
+export type Indicator = {
+  column: string;
+  name: string;
+  value: string[];
+  status: IndicatorStatus;
+  path: string[];
+};
+
+export const selectIndicatorsForChart = (
+  chartId: number,
+  filters: { [key: number]: Filter },
+  datasources: { [key: string]: Datasource },
+  charts: any,
+) => {
+  const chart = charts[chartId];
+  // for now we only need to know which columns are compatible/incompatible,
+  // so grab the columns from the applied/rejected filters
+  const appliedColumns: Set<string> = new Set(
+    get(chart, 'queryResponse.applied_filters', []).map(
```

Comment:
This can probably just use safe navigation operators (`this?.that?.thing`) to skip the need for lodash's `get` in this file